pkg/raft/term_cache_test.go

```diff
@@ -0,0 +1,505 @@
+		require.Equal(t, tcase.wantOK, ok, "index %d", tcase.index)
+	}
+}
+
+func TestTermCachePopulatedTruncateAndAppend(t *testing.T) {
+	testCases := []struct {
+		name            string
+		cacheSize       uint64
+		initialID       entryID
+		initTerms       []uint64 // Initial terms to populate cache
+		appendInitialID entryID
+		appendTerms     []uint64 // Terms to append after initial population
+		want            struct {
+			cacheLen             int
+			first                entryID
+			last                 entryID
+			indexAndExpectedTerm [][2]uint64 // Checks to verify terms at specific indices
+		}
+	}{
+		{
+			name:      "append to half-filled cache",
+			cacheSize: 4,
+			initialID: entryID{term: 5, index: 10},
+			// Initial append fills the cache:
+			// [t5/10, t6/12, t6/13]
+			initTerms:       []uint64{5, 6, 6},
+			appendInitialID: entryID{term: 6, index: 13},
+			// Second append:
+			// [t7/14, t8/15]
+			appendTerms: []uint64{7, 8},
+			want: struct {
+				cacheLen             int
+				first                entryID
+				last                 entryID
+				indexAndExpectedTerm [][2]uint64
+			}{
+				cacheLen: 4,
+				first:    entryID{term: 5, index: 10}, // First entry
+				last:     entryID{term: 8, index: 15}, // New last entry
+			},
+		},
+		{
+			name:      "cause eviction, no overwrite",
+			cacheSize: 4,
+			initialID: entryID{term: 5, index: 10},
+			// Initial entries only fill half: [t5/10, t6/11]
+			initTerms:       []uint64{6},
+			appendInitialID: entryID{term: 6, index: 11},
+			// Append more: [t7/12, t8/13, t9/14]
+			appendTerms: []uint64{7, 8, 9},
+			// Final term cache state: [t6/11, t7/12, t8/13, t9/14]
+			want: struct {
+				cacheLen             int
+				first                entryID
+				last                 entryID
+				indexAndExpectedTerm [][2]uint64
+			}{
+				cacheLen: 4,                           // Now full
+				first:    entryID{term: 6, index: 11}, // Nothing evicted yet
+				last:     entryID{term: 9, index: 14}, // New last entry
+			},
+		},
+		{
+			name:      "overwriting with new entries",
+			cacheSize: 4,
+			initialID: entryID{term: 5, index: 10},
+			// Initial entries: [t5/10, t6/11, t7/12, t8/13], already full
+			initTerms:       []uint64{6, 7, 8},
+			appendInitialID: entryID{term: 5, index: 10},
+			// Many term flips: [t11/11, t12/12, t13/13]
+			// We want to overwrite the original termcache, starting at t6/11
+			appendTerms: []uint64{11, 12, 13},
+			// Final term cache state: [t5/10, t11/11, t12/12, t13/13]
+			want: struct {
+				cacheLen             int
+				first                entryID
+				last                 entryID
+				indexAndExpectedTerm [][2]uint64
+			}{
+				cacheLen: 4,                            // At maximum
+				first:    entryID{term: 5, index: 10},  // First entry evicted
+				last:     entryID{term: 13, index: 13}, // New last entry
+				indexAndExpectedTerm: [][2]uint64{
+					{10, 5},
+					{11, 11},
+					{12, 12},
+					{13, 13},
+				},
```

Comment:
This trio could be squashed into being a `LogSlice`:

```golang
	want: entryID{term: 5, index: 10}.append(10, 11, 12),
```

Then, the test needs to make sure 3 things:

1. `term()` for indices below `prev.index` of this slice return `false`.
2. Indices within the `LogSlice` bounds return the correct term == `LogSlice.termAt(index)`.
3. Indices above the `lastIndex` return the last term.